File: backend/internal/handler/category.go
package handler

import (
	"net/http"

	"github.com/vibestore/backend/internal/dto"
	"github.com/vibestore/backend/internal/middleware"
	"github.com/vibestore/backend/internal/model"
	"github.com/vibestore/backend/internal/repository"
)

type CategoryHandler struct {
	categoryRepo *repository.CategoryRepository
}

func NewCategoryHandler(categoryRepo *repository.CategoryRepository) *CategoryHandler {
	return &CategoryHandler{categoryRepo: categoryRepo}
}

// GET /api/v1/store/categories - Public, returns tree
func (h *CategoryHandler) ListPublic(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetStoreID(r.Context())
	categories, err := h.categoryRepo.FindActiveByStoreID(r.Context(), storeID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to list categories")
		return
	}
	tree := repository.BuildCategoryTree(categories)
	writeJSON(w, http.StatusOK, tree)
}

// GET /api/v1/admin/categories - Admin, returns flat list
func (h *CategoryHandler) List(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	categories, err := h.categoryRepo.FindByStoreID(r.Context(), storeID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to list categories")
		return
	}
	if categories == nil {
		categories = []model.Category{}
	}
	writeJSON(w, http.StatusOK, categories)
}

// GET /api/v1/admin/categories/{id}
func (h *CategoryHandler) Get(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	id, err := urlParamUint64(r, "id")
	if err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid category ID")
		return
	}
	category, err := h.categoryRepo.FindByID(r.Context(), id, storeID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to get category")
		return
	}
	if category == nil {
		writeError(w, http.StatusNotFound, "NOT_FOUND", "Category not found")
		return
	}
	writeJSON(w, http.StatusOK, category)
}

// POST /api/v1/admin/categories
func (h *CategoryHandler) Create(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())

	var req dto.CategoryRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid request body")
		return
	}
	if req.Name == "" || req.Slug == "" {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Name and slug are required")
		return
	}

	category := &model.Category{
		StoreID:     storeID,
		ParentID:    req.ParentID,
		Name:        req.Name,
		Slug:        req.Slug,
		Description: req.Description,
		ImageURL:    req.ImageURL,
		SortOrder:   req.SortOrder,
		IsActive:    req.IsActive,
	}

	if err := h.categoryRepo.Create(r.Context(), category); err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to create category")
		return
	}

	writeJSON(w, http.StatusCreated, category)
}

// PUT /api/v1/admin/categories/{id}
func (h *CategoryHandler) Update(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	id, err := urlParamUint64(r, "id")
	if err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid category ID")
		return
	}

	existing, err := h.categoryRepo.FindByID(r.Context(), id, storeID)
	if err != nil || existing == nil {
		writeError(w, http.StatusNotFound, "NOT_FOUND", "Category not found")
		return
	}

	var req dto.CategoryRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid request body")
		return
	}

	existing.ParentID = req.ParentID
	existing.Name = req.Name
	existing.Slug = req.Slug
	existing.Description = req.Description
	existing.ImageURL = req.ImageURL
	existing.SortOrder = req.SortOrder
	existing.IsActive = req.IsActive

	if err := h.categoryRepo.Update(r.Context(), existing); err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to update category")
		return
	}

	writeJSON(w, http.StatusOK, existing)
}

// DELETE /api/v1/admin/categories/{id}
func (h *CategoryHandler) Delete(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	id, err := urlParamUint64(r, "id")
	if err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid category ID")
		return
	}

	if err := h.categoryRepo.Delete(r.Context(), id, storeID); err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to delete category")
		return
	}

	writeJSON(w, http.StatusOK, map[string]string{"message": "Category deleted"})
}

File: backend-bun/src/routes/category.ts
import Elysia from "elysia";
import { tenantMiddleware } from "../middleware/tenant";
import { authMiddleware } from "../middleware/auth";
import * as catRepo from "../repositories/category";
import type { CategoryRequest } from "../types";

export const publicCategoryRoutes = new Elysia({ prefix: "/api/v1" })
  .use(tenantMiddleware)
  .get("/store/categories", async ({ storeId }) => {
    const cats = await catRepo.findActiveCategoriesByStoreID(storeId);
    return catRepo.buildCategoryTree(cats);
  });

export const adminCategoryRoutes = new Elysia({ prefix: "/api/v1/admin/categories" })
  .use(tenantMiddleware)
  .use(authMiddleware)
  .get("/", async ({ storeId }) => {
    return await catRepo.findCategoriesByStoreID(storeId);
  })
  .get("/:id", async ({ params, storeId, set }) => {
    const cat = await catRepo.findCategoryByID(Number(params.id), storeId);
    if (!cat) {
      set.status = 404;
      return { code: "NOT_FOUND", message: "Category not found" };
    }
    return cat;
  })
  .post("/", async ({ body, storeId, set }) => {
    const req = body as CategoryRequest;
    const id = await catRepo.createCategory({ ...req, store_id: storeId });
    const cat = await catRepo.findCategoryByID(id, storeId);
    set.status = 201;
    return cat;
  })
  .put("/:id", async ({ params, body, storeId, set }) => {
    const existing = await catRepo.findCategoryByID(Number(params.id), storeId);
    if (!existing) {
      set.status = 404;
      return { code: "NOT_FOUND", message: "Category not found" };
    }
    const req = body as CategoryRequest;
    const updated = {
      ...existing,
      name: req.name,
      slug: req.slug,
      parent_id: req.parent_id ?? null,
      description: req.description ?? null,
      image_url: req.image_url ?? null,
      sort_order: req.sort_order,
      is_active: req.is_active,
    };
    await catRepo.updateCategory(updated);
    return await catRepo.findCategoryByID(existing.id, storeId);
  })
  .delete("/:id", async ({ params, storeId, set }) => {
    const existing = await catRepo.findCategoryByID(Number(params.id), storeId);
    if (!existing) {
      set.status = 404;
      return { code: "NOT_FOUND", message: "Category not found" };
    }
    await catRepo.deleteCategory(existing.id, storeId);
    set.status = 204;
  });
